base/range.jl

```diff
@@ -871,13 +871,22 @@ end
 
 median(r::Range{<:Real}) = mean(r)
 
-function in(x, r::Range)
-    n = step(r) == 0 ? 1 : round(Integer,(x-first(r))/step(r))+1
-    n >= 1 && n <= length(r) && r[n] == x
+function _in_range(x, r::Range)
+    if step(r) == 0
+        return !isempty(r) && first(r) == x
+    else
+        n = round(Integer, (x - first(r)) / step(r)) + 1
+        return n >= 1 && n <= length(r) && r[n] == x
+    end
 end
+in(x::Real, r::Range{<:Real}) = _in_range(x, r)
+# This method needs to be defined separately since -(::T, ::T) can be implemented
+# even if -(::T, ::Real) is not
+in(x::T, r::Range{T}) where {T} = _in_range(x, r)
 
 in(x::Integer, r::AbstractUnitRange{<:Integer}) = (first(r) <= x) & (x <= last(r))
-in(x, r::Range{T}) where {T<:Integer} =
```

Comment:
It's more restrictive in the sense that the fast path is only used for types for which we're it will work (i.e. `Real`, for which `isinteger` and arithmetic operations must be defined). But it's not more restrictive in general thanks to the `in(x, itr)` fallback. As noted in the description, previously checking that a non-numeric type was `in` an integer range would fail, which is quite "restrictive" (e.g. `"a" in 1:10` wouldn't work).